tuf/repository_lib.py

```diff
@@ -1466,6 +1431,47 @@ def generate_targets_metadata(targets_directory, target_files, version,
 
 
 
+def _add_keys_to_keydict(keydict, keyids, repository_name='default'):
+  """
+  Non-public function that generates keys for the keyids listed and adds
+  them to a key dictionary.
+  """
+
+  for keyid in keyids:
+
+    # Do we have a duplicate?
+    if keyids.count(keyid) > 1:
+      raise securesystemslib.exceptions.Error('Same keyid listed twice:'
+          ' ' + keyid)
+
+    key = tuf.keydb.get_key(keyid, repository_name=repository_name)
+
+    # If 'key' is an RSA key, it would conform to
+    # 'securesystemslib.formats.RSAKEY_SCHEMA', and have the form:
+    # {'keytype': 'rsa',
+    #  'keyid': keyid,
+    #  'keyval': {'public': '-----BEGIN RSA PUBLIC KEY----- ...',
+    #             'private': '-----BEGIN RSA PRIVATE KEY----- ...'}}
+    keyid = key['keyid']
+    if keyid not in keydict:
+
+      # This appears to be a new keyid.  Generate the key for it.
+      if key['keytype'] in ['rsa', 'ed25519', 'ecdsa-sha2-nistp256']:
```

Comment:
I wonder if there's a reason for hard coding these here: we could just let format_keyval_to_metadata() throw FormatError (or handle it and throw something better) ?

Reply:
Thanks for pointing it :) 
After a quick search,  `format_keyval_to_metadata()` is used at a couple of places in `tuf` the way you described it (just letting it throw FormatError) so I agree applying the same here.

It seems also worth trying to merge the functionality of `repository_tool._keys_to_keydict()` and `repository_lib._add_keys_to_keydict()` which seem suspiciously similar :eyes: